File: Best_Research/#Done/Rust/LearningRust/Units/Units/src/units/time.rs
use std::fmt;

use crate::internal::{
    QuantityError, format_unit_value, implement_quantity_arithmetic, validate_finite,
};

#[repr(transparent)]
#[derive(Copy, Clone, PartialEq, PartialOrd, Debug, Default)]
pub struct Time(pub(crate) f64);

impl Time {
    #[must_use]
    pub const fn seconds(value: f64) -> Self {
        Self(value)
    }

    #[must_use]
    pub const fn milliseconds(value: f64) -> Self {
        Self(value / 1_000.0)
    }

    #[must_use]
    pub const fn microseconds(value: f64) -> Self {
        Self(value / 1_000_000.0)
    }

    #[must_use]
    pub const fn nanoseconds(value: f64) -> Self {
        Self(value / 1_000_000_000.0)
    }

    #[must_use]
    pub const fn minutes(value: f64) -> Self {
        Self(value * 60.0)
    }

    #[must_use]
    pub const fn hours(value: f64) -> Self {
        Self(value * 3_600.0)
    }

    pub fn try_seconds(value: f64) -> Result<Self, QuantityError> {
        Ok(Self(validate_finite("Time", "s", value)?))
    }

    #[must_use]
    pub const fn as_seconds(self) -> f64 {
        self.0
    }

    #[must_use]
    pub const fn as_milliseconds(self) -> f64 {
        self.0 * 1_000.0
    }

    #[must_use]
    pub const fn as_microseconds(self) -> f64 {
        self.0 * 1_000_000.0
    }

    #[must_use]
    pub const fn as_nanoseconds(self) -> f64 {
        self.0 * 1_000_000_000.0
    }

    #[must_use]
    pub const fn as_minutes(self) -> f64 {
        self.0 / 60.0
    }

    #[must_use]
    pub const fn as_hours(self) -> f64 {
        self.0 / 3_600.0
    }

    #[must_use]
    pub const fn display_as(self, unit: TimeUnit) -> TimeDisplay {
        TimeDisplay {
            value: self,
            unit,
            precision: None,
        }
    }

    #[must_use]
    pub const fn display_as_precision(self, unit: TimeUnit, precision: usize) -> TimeDisplay {
        TimeDisplay {
            value: self,
            unit,
            precision: Some(precision),
        }
    }
}

#[must_use]
pub const fn seconds(value: f64) -> Time {
    Time::seconds(value)
}

#[must_use]
pub const fn milliseconds(value: f64) -> Time {
    Time::milliseconds(value)
}

#[must_use]
pub const fn microseconds(value: f64) -> Time {
    Time::microseconds(value)
}

#[must_use]
pub const fn nanoseconds(value: f64) -> Time {
    Time::nanoseconds(value)
}

#[must_use]
pub const fn minutes(value: f64) -> Time {
    Time::minutes(value)
}

#[must_use]
pub const fn hours(value: f64) -> Time {
    Time::hours(value)
}

implement_quantity_arithmetic!(Time);

#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub enum TimeUnit {
    Seconds,
    Milliseconds,
    Microseconds,
    Nanoseconds,
    Minutes,
    Hours,
}

impl TimeUnit {
    #[must_use]
    pub const fn symbol(self) -> &'static str {
        match self {
            Self::Seconds => "s",
            Self::Milliseconds => "ms",
            Self::Microseconds => "us",
            Self::Nanoseconds => "ns",
            Self::Minutes => "min",
            Self::Hours => "h",
        }
    }
}

#[derive(Copy, Clone, Debug)]
pub struct TimeDisplay {
    value: Time,
    unit: TimeUnit,
    precision: Option<usize>,
}

impl fmt::Display for TimeDisplay {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = match self.unit {
            TimeUnit::Seconds => self.value.as_seconds(),
            TimeUnit::Milliseconds => self.value.as_milliseconds(),
            TimeUnit::Microseconds => self.value.as_microseconds(),
            TimeUnit::Nanoseconds => self.value.as_nanoseconds(),
            TimeUnit::Minutes => self.value.as_minutes(),
            TimeUnit::Hours => self.value.as_hours(),
        };

        format_unit_value(formatter, value, self.unit.symbol(), self.precision)
    }
}

impl fmt::Display for Time {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.display_as(TimeUnit::Seconds).fmt(formatter)
    }
}

File: Best_Research/#Done/C++/Units-CppXmake/Physics/time/time.cpp
#include "Physics/time/time.hpp"

#include "Physics/detail/math.hpp"

namespace Physics::time
{

    Time::Time(double seconds) : _seconds(seconds)
    {
    }

    Time Time::fromRawSi(double seconds)
    {
        return Time(seconds);
    }

    Time Time::seconds(double value)
    {
        return Time(value);
    }

    Time Time::milliseconds(double value)
    {
        return Time(value / 1'000.0);
    }

    Time Time::microseconds(double value)
    {
        return Time(value / 1'000'000.0);
    }

    Time Time::nanoseconds(double value)
    {
        return Time(value / 1'000'000'000.0);
    }

    Time Time::minutes(double value)
    {
        return Time(value * 60.0);
    }

    Time Time::hours(double value)
    {
        return Time(value * 3'600.0);
    }

    double Time::rawSi()
    {
        return _seconds;
    }

    double Time::asSeconds()
    {
        return _seconds;
    }

    double Time::asMilliseconds()
    {
        return _seconds * 1'000.0;
    }

    double Time::asMicroseconds()
    {
        return _seconds * 1'000'000.0;
    }

    double Time::asNanoseconds()
    {
        return _seconds * 1'000'000'000.0;
    }

    double Time::asMinutes()
    {
        return _seconds / 60.0;
    }

    double Time::asHours()
    {
        return _seconds / 3'600.0;
    }

    bool Time::approximatelyEquals(Time other, double epsilon)
    {
        return detail::absolute(_seconds - other._seconds) <= epsilon;
    }

    QuantityDisplay<Time, TimeUnit> Time::displayAs(TimeUnit unit)
    {
        return QuantityDisplay<Time, TimeUnit>(*this, unit);
    }

    QuantityDisplay<Time, TimeUnit> Time::displayAsPrecision(TimeUnit unit, int precision)
    {
        return QuantityDisplay<Time, TimeUnit>(*this, unit, precision);
    }

    bool operator==(Time left, Time right)
    {
        return left._seconds == right._seconds;
    }

} // namespace Physics::time
